src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,318 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Load, update, monitor and retrieve machine learning models."""
+
+import abc
+import asyncio
+import logging
+import os
+import pickle
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Generic, TypeVar
+
+from frequenz.channels.util import FileWatcher
+
+from frequenz.sdk._internal._asyncio import cancel_and_await
+
+_logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+"""The model type."""
+
+
+class ModelRepository(abc.ABC):
+    """Provide an interface for building the model repository configuration.
+
+    The ModelRepository is meant to be subclassed to provide specific
+    implementations based on the type of models being managed.
+    """
+
```

Comment:
Initially we discussed to have `allowed_keys` as a field in this base class. By this you would not need the `_get_allowed_keys` method. What was the problem with that again?

Reply:
I think the main reason is if we do that then the `ModelRepository` does not make any sense. I'll elaborate on that with the following question, what would it be the difference between passing a dictionary {key: path} as it was previously done and passing a `ModelRepository` subclass with `allowed_keys`, `directory`, `model`, etc?
The idea of the `ModelRepository` interface was to encourage the user to use the pre-defined `SingleModelRepository` or `EnumeratedRepository` where the keys have been already defined to avoid and somehow "restrict" them to generate their own arbitrary keys. 

By defining the abstract methods `_get_allowed_keys()` and `_build_path()`, the user can customize keys/paths in case they really needed for edge or corner cases and this somehow will make them aware that if they want to define any kind of arbitrary keys or paths then they need a specialization of the `ModelRepository`

Reply:
I don't see much difference in writing
```
@dataclass
class WeekdayModelRepository(ModelRepository):
    def _get_allowed_keys(self) -> list[Any]:
        return ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
...
```
vs 
```
@dataclass
class WeekdayModelRepository(ModelRepository):
    allowed_keys = ("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")
...
```
Personally I find latter clearer. 

If you want to prevent the user to change keys, there is a frozen option for dataclasses. If you are concerned about types, we could think of requiring dedicated `Key` types. 

I agree that ideally users are encouraged to use the pre-defined implementations, and I think those cover most of our cases already. But to me the most striking point was the `_build_path` method which encourages to structure your folders in line with the keys.

Reply:
> Personally I find latter clearer

I agree the attribute `allowed_key` is clearer but that will allow the user to inject any kind of key if there is no a frozen technique or interface in place (e.g. the `EnumeratedModelRepository` will no longer be enumerated if the user injects any arbitrary key).

> If you want to prevent the user to change keys, there is a frozen option for dataclasses. If you are concerned about types, we could think of requiring dedicated Key types.

I think the `frozen` option will apply for all attributes including the base directory and model name, so if we go that way we might need to provide a `classmethod` decorator to instantiate the model repository. 

> I agree that ideally users are encouraged to use the pre-defined implementations, and I think those cover most of our cases already. But to me the most striking point was the _build_path method which encourages to structure your folders in line with the keys.

I understood  (probably I misunderstood) the folder structure includes the use of the predefined keys. If this is not the case then we should go with the attribute `allowed_keys`. I think this the point we need to clarify and agree on.

Reply:
I agree that

```py
@dataclass
class WeekdayModelRepository(ModelRepository):
    allowed_keys = ("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")
...
```
looks nicer but then the user could overwrite `allowed_key`, which is not good. I prefer to be on the safe (or clear API) side and keep it as a method, to make it very clear that only the creator of the model repository can decide which keys are allowed. And yeah, marking the whole dataclass frozen is not super practical. I also think it is a feature to make creating new model repositories non-trivial, as it should be discouraged.

@cwasicki what is your main concern about keeping the `_get_allowed_keys` abstract method?